cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt

```diff
@@ -0,0 +1,262 @@
+    private val accessQueue: MutableSet<CacheEntry<Key, Value>>?
+
+    /**
+     * Whether to perform size based evictions.
+     */
+    private val evictsBySize = maxSize >= 0
+
+    /**
+     * Whether to perform write-time based expiration.
+     */
+    private val expiresAfterWrite = expireAfterWriteNanos > 0
+
+    /**
+     * Whether to perform access-time (both read and write) based expiration.
+     */
+    private val expiresAfterAccess = expireAfterAccessNanos > 0
+
+    /**
+     * A key-based synchronizer for running cache loader.
+     */
+    private val loaderSynchronizer = KeyedSynchronizer<Key>()
+
+    init {
+        // writeQueue is required if write expiry is enabled
+        writeQueue = takeIf { expiresAfterWrite }?.let {
+            Collections.synchronizedSet(ReorderingLinkedHashSet())
+        }
+
+        // accessQueue is required if either read expiry is enabled or size-based eviction is enabled
+        accessQueue = takeIf { expiresAfterAccess || evictsBySize }?.let {
+            Collections.synchronizedSet(ReorderingLinkedHashSet())
+        }
+    }
+
+    override fun get(key: Key): Value? {
+        val nowNanos = clock.currentTimeNanos
+        return cacheEntries[key]?.let {
+            if (it.isExpired(nowNanos)) {
+                // clean up expired entries and return null
+                expireEntries(nowNanos)
+                null
+            } else {
+                // update eviction metadata and return the value
+                recordRead(it, nowNanos)
+                it.value
+            }
+        }
+    }
+
+    override fun get(key: Key, loader: () -> Value): Value {
+        return loaderSynchronizer.run(key) {
+            val nowNanos = clock.currentTimeNanos
+            cacheEntries[key]?.let {
+                if (it.isExpired(nowNanos)) {
+                    // clean up expired entries
+                    expireEntries(nowNanos)
+                    null
+                } else {
+                    // update eviction metadata
+                    recordRead(it, nowNanos)
+                    it.value
+                }
+            } ?: loader().let { loadedValue ->
+                // cache the loaded value if value for the key is still absent
+                // from the cache entries after executing the loader
+                val existingValue = get(key)
+                if (existingValue != null) {
+                    existingValue
+                } else {
+                    put(key, loadedValue)
+                    loadedValue
+                }
+            }
+        }
+    }
+
+    override fun put(key: Key, value: Value) {
+        val nowNanos = clock.currentTimeNanos
+        expireEntries(nowNanos)
+
+        val existingEntry = cacheEntries[key]
+        if (existingEntry != null) {
+            // cache entry found
+            recordWrite(existingEntry, nowNanos)
+            existingEntry.value = value
+            cacheEntries[key] = existingEntry
```

Comment:
can you add a comment about memory barrier to explain for future readers why this line is necessary